### PR TITLE
ASSOCIATIONS RoR

### PR DESCRIPTION
**Pull request by Rocio01**
Greetings
This a pull request for the project Associations(RoR) We built a site similar to a private Eventbrite which allows users to create events.
Kind  Regards

### CODE REVIEW

File: app/views/events/_form.html.erb
Comment (teekaytech) [inline app/views/events/_form.html.erb]: There should be no application logic in your views. 
Views are only for displaying information that became available to it from its corresponding controller not generating it.

The following resources might help:
- [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
- [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
- [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright

File: app/views/users/show.html.erb
Comment (teekaytech) [inline app/views/users/show.html.erb]: This data fetching `current_user.events` should be called in your `users_controller#show`, just like the way you did it in your `events_controller#index action.

File: app/views/users/show.html.erb
Comment (teekaytech) [inline app/views/users/show.html.erb]: This also needs to be refactored. Fetch the data from the controller.

**Review by teekaytech — CHANGES_REQUESTED** [review]
## Status: Changes Required :warning: :vertical_traffic_light: 

### Hello team :wave:

__You've made laudable progress in this project. Your test suites are working well and your code base is clean. :+1: 
However, some changes are required before this project can be approved. Please check my comments below and fix them as appropriate.__

**[ REQUIRED CHANGES ]**

- [ ] I ran into series of challenges while trying to run your project on my PC. 
- The first was about your Ruby version. In your `.ruby-version` file, you specified 2.7.1 while in your gemfile, you specified 2.7.2. Use a consistent ruby version.
- The second challenge has to do with one of your gems (WDM). You should know that this gem is useful on Windows platform. As a result, you should try to include the necessary conditions on how to ignore the gem for other platforms (Linux, etc). [This SOF thread](https://stackoverflow.com/questions/16232960/guard-wont-load-wdm) might help. Here's the error message:


- [ ] The user's show page should display events the user has created and events that the user is attending to separated in past and upcoming ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
- Right now on your app, a logged in user can see `created events` and `attended events`. However, these 2 categories of events should be sub-divided into past and upcoming events.

- [ ] Everything connected with data processing should be encapsulated in models and used in controllers.  
- i.e. no `Post.new`, `Comment.create`, `current_user.friends.find_by(...)`...).
- Check my comments below for details.

- [ ] Files not related to the project should be completely removed and added to .gitignore. Delete the `.browserslistrc` and add it to your `.gitignore` file.



* * *
**Please make the required changes and submit another review request when you are done.**

* * *

**Taofeek Olalere
[Twitter](https://twitter.com/ola_lere) | [Github](https://github.com/teekaytech) | @taofeekolalere on Slack**
__Happy Coding :smiley:__

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

Hello @karthykarthick and @Rocio01, Great job! 👏  Both of you have done well on this project. Amazing improvement from the last review.

However, there are still some changes that are necessary for this project to be approved.

Kindly take a look at these view files with logic in them:
- events/_form.html.erb
- events/index.html.erb
- layouts/application.html.erb
- users/_form.html.erb
- users/show.html.erb

As the last reviewer pointed out, there should be absolutely NO logic in views. >Views are only for displaying information that became available to it from its corresponding controller not generating it.  **The Resources below could be of immense help**
* [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
* [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
* [Rails View Helpers Best Practices](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4)
* [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright

- [ ] Currently, a user can modify and delete an event he/she did not create. Kindly fix this and make sure you can only modify or delete an event you created.

- [ ] For some reasons, I can not attend upcoming events, Kindly fix this as well.

- [ ] Currently, a user can attend an event as much time as he clicks attend the event. You should be able to attend an event just once.
Screen Shot 2020-12-09 at 6 55 23 PM


### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

File: app/views/users/_form.html.erb
Comment (ghost) [inline app/views/users/_form.html.erb]: I can't find where you have referenced this partial. Delete it since it is of no use.

**Review by ghost — CHANGES_REQUESTED** [review]
# Status: Changes Required 🤓

## Hello team 😄,

**Great job implementing the requested changes. However, there are still a few fixes required.**

1). **Your views folder still has logic in form of conditional statements and loops. View-related logic can be encapsulated in `View helpers (app/helpers/*_helper.rb)` and not `partial templates` as you have done. Hint: You can check [this](https://guides.rubyonrails.org/v5.0/layouts_and_rendering.html) guideline on how to use rendering collections for loops. The view files with logic include:**

- all the partial templates in the `events` directory.
- `layouts/_header.html.erb`.
- `users/_show.html.erb`.

2). **I will use an example. Jen has created an event for Dec 13th. When Ken clicks to attend the event, it doesn't show on his list of upcoming events on his profile page. Also, Jen created another event for Dec 1st, when Ken clicks to attend, it doesn't show on his list of previous events.**

3).**As mentioned by the previous TSE, fix attend event button so the user is not able to click it more than once.**

4). **You have a failing RSpec test.**
rspec

5). **Fix the linter errors.**


## Happy coding 🥂.

**PS: You can reach me on slack (Rose Wanjohi).**

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

### Hello Team!

You've done an awesome job so far 👌. But there are still some changes that are necessary for this project to be approved.

#### Kindly take a look at these view files, you still have logic in them:
- events/index.html.erb
- layouts/application.html.erb
- users/show.html.erb

There should be absolutely NO logic in views. >Views are only for displaying information that became available to it from its corresponding controller not generating it.  **The Resources below could be of immense help**
* [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
* [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
* [Rails View Helpers Best Practices](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4)
* [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright

- [x] Kindly add outline/border to your input fields as it's really hard to know where to type. 
Screen Shot 2020-12-11 at 5 40 23 PM
Screen Shot 2020-12-11 at 6 04 47 PM

- [x] Good test coverage, just remove all pending examples 

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

**Comment by Rocio01** [comment]
Greetings
 We face a problem with the logic in the views because they have many html code in the html.erb files inside the logic and we can't find a way to translate that in the helper methods.
Thanks for the attention
Kind regards
An example 
ejemplo

File: app/views/events/index.html.erb
Comment (kevcap) [inline app/views/events/index.html.erb:6]: ❌ Please remove inline styling;

File: app/views/events/index.html.erb
Comment (kevcap) [inline app/views/events/index.html.erb:25]: ❌ Please remove inline styling;

File: app/views/events/show.html.erb
Comment (kevcap) [inline app/views/events/show.html.erb:19]: ❌ Please remove inline styling;

File: app/views/events/show.html.erb
Comment (kevcap) [inline app/views/events/show.html.erb:36]: ❌ Please remove inline styling;

**Review by kevcap — CHANGES_REQUESTED** [review]
# [STATUS] CHANGES REQUIRED♻️

#### Hey team,  your project is looking really good, however, to accomplish as the actual project, you required to do these changes, listed below: 
 
## [REQUIRED CHANGES]
- [x] Please don't use inline styling;
- [x] You should only show the events if the user has logged in;
Currenly this is your homepage:
image
- [ ] When we click on show event, it throws an reference error, please fix it;
- [ ] You have to validate the input data!! currently users can type text on date fields! ⚠
image
- [ ] We have rspec errors, please fix them;

## Please make the required changes and resubmit for a new code review once you are ready💪

## As always, I wish you all the very best and see you both at the top🚀


👨🏻‍💻 **TSE: Kevin Alves**

  

📫 Have any question? Better call Kevin:

<a  href="mailto:kevinalves.js@gmail.com"> gmail</a> [Twitter Follow](https://twitter.com/intent/follow?screen_name=kevinnnmn) [](https://microverse-students.slack.com/team/UT86D3AP4)

<sub>*We're here to help you, to ensure that you're using the best practices and getting you ready for the job market. In real-world, you'll have a lot of changes requests for your real clients, it's good to get used to it.</sub>

**Comment by Rocio01** [comment]
Greetings
Thanks for the review
On the original Eventbrite page you can see the event even if you are not logged in, so we added a home page, but you can still view the events without logging in. Only attend when you are logged in.
Kind regards.

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello Zulma Rocio Martinez Guarin & Karthick Harimoorthy** 
Nice job so far :muscle:  :muscle:
Your project meets most of the requirements :clap: :clap: :clap:  
Still, there is some improvement to do in your tests before you can go ahead:

- [ ] Complete pending tests or remove them
- [ ] In your tests, each scenario should only have one expect
- [ ] Remove commented code from the files in spec/helpers folder
- [ ] Remove commented code in event_attendence_spec.rb file
- [ ] Remove spec files with no tests
 
Please request another review once you are ready! :+1: 
 
Happy coding :relaxed:
 
Oscar Andrés Russi Porras

**Review by Temmarie — CHANGES_REQUESTED** [review]
## Changes Required! :recycle:
### Hi @Rocio01 and @karthykarthick ! 👍 👋 👋 

### You have done well so far :clap: However, you did not finish all the previous review changes. Kindly work on them and submit them for another code review.

**Recommended**

- [ ] Delete empty test files.

- [ ] FIx the errors in your test files.

Screenshot (156)



__Kindly make the necessary changes and submit another code review request.__
_Feel free to reach me on slack for any inquiries about this review._ :wink:
__Grace Tamara Ekunola __
[GitHub](https://github.com/temmarie) | [LinkedIn](https://www.linkedin.com/in/ekunola-grace/) | [Twitter](https://twitter.com/TemmarieW)

**Comment by Rocio01** [comment]
Greetings
Thank you for your revision, but on our computers, we don't get errors in the test files.
Kind regards
spec

**Review by javitocor — APPROVED** [review]
# STATUS APPROVED 🥇 

Hello team! ✋ 
Your project meets all the basic requirements! ⭐ 
Congratulations! 🥳  It has been approved! 👍 
Remember to fill up the completion form 📑 , merge it and you can move into the next project 🚀 
Good luck with it! 😺

**Pull request merged**
